raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java

```diff
@@ -0,0 +1,377 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.raft.internals;
+
+import java.net.InetSocketAddress;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.feature.SupportedVersionRange;
+import org.apache.kafka.common.message.VotersRecord;
+import org.apache.kafka.common.utils.Utils;
+
+/**
+ * A type for representing the set of voters for a topic partition.
+ *
+ * It encapsulates static information like a voter's endpoint and their supported kraft.version.
+ *
+ * It providees functionality for converting to and from {@code VotersRecord} and for converting
+ * from the static configuration.
+ */
+final public class VoterSet {
+    private final Map<Integer, VoterNode> voters;
```

Comment:
Yes. That is a sufficient condition but I decided to implement a stricter condition that the ids need be unique. This means that with this implementation, if they want to replace a directory id, they need to first remove the failed replica key (id, uuid) and then add the new replica key (id, uuid').

There are two reasons why I decide to keep the ids unique:
1. It makes it easier and safer to implement the feature for automatically having new controllers join the quorum. I was concerned that the set of voters would become unavailable if there was a race where new directory id kept joining the cluster. In this example the cluster would be come unavailable and the user would not be able to mitigate it: [(1, uuid1), (2, uuid2), (3, uuid3), (3, uuid3'), (3, uuid3''), (3, uuid3''')].
2. Connection management is easier to implement. A lot of code in Kafka (e.g. `o.a.k.c.Node`) assumes that ids are unique and they can be used, along with the listener name, to identify an endpoint. I think it would be a big effort to extend this to identify endpoint by the replica key. Another example is `NodeEndpoints` in `FetchResponse`. That map is index by replica id.

The main disadvantage of this implementation is that it would make it difficult to design and implement the ability for dynamically "altering the metadata/kraft log directory" like Kafka does for regular topic partitions. But I think we can discuss that if we ever want to implement that feature in the future.

Reply:
This approach is fine, but we need to update the KIP.

Reply:
I think I already updated the KIP to match this comment but I'll take a look again.